### DISCUSSION TITLE
Players get negative claim blocks after buying some more claim blocks

### DISCUSSION THREAD

**Discussion by Compositr** [GriefPrevention/GriefPrevention#863] [Q&A] [answered]
Note: I'm running 1.16.1, if that affects anything

### What happened:
_Some players on my server got their claim blocks balance reset to negative numbers, usually after attempting to buy more claim blocks using /buyclaimblocks_

### What was expected:
_The appropriate amount of claim blocks added to their balance_

### Steps to reproduce:
1. Buy some claim blocks using an eco system (I use EssentialsEco, don't know if that affects it)
2. Grab the golden shovel to check how many blocks you have
3. You get a negative amount of claim blocks

### Server and GriefPrevention version:
/version 
`[INFO] This server is running CraftBukkit version git-Spigot-ad703da-d77f4d9 (MC: 1.16.1) (Implementing API version 1.16.1-R0.1-SNAPSHOT)

[INFO] Checking version, please wait...

[INFO] You are 4 version(s) behind`
/version griefprevention `GriefPrevention version 16.13.0`: 

### Stack trace/error or server log

```
The console doesn't seem to have any error showing.

```
Or, paste the server.log at gist.github.com and paste the link here: 

### GriefPrevention config.yml

```
# Default values are perfect for most servers.  If you want to customize and have a question, look for the answer here first: http://dev.bukkit.org/bukkit-plugins/grief-prevention/pages/setup-and-configuration/
GriefPrevention:
  SeaLevelOverrides:
    world: -1
    world_nether: -1
    world_the_end: -1
  Claims:
    Mode:
      world_the_end: Disabled
      world: Survival
      world_nether: Disabled
    PreventGlobalMonsterEggs: true
    PreventTheft: true
    ProtectCreatures: true
    PreventButtonsSwitches: true
    LockWoodenDoors: false
    LockTrapDoors: false
    LockFenceGates: true
    EnderPearlsRequireAccessTrust: true
    ProtectHorses: true
    ProtectDonkeys: true
    ProtectLlamas: true
    InitialBlocks: 100
    Claim Blocks Accrued Per Hour:
      Default: 50
    Max Accrued Claim Blocks:
      Default: 5000
    Accrued Idle Threshold: 0
    AccruedIdlePercent: 0
    AbandonReturnRatio: 1.0
    AutomaticNewPlayerClaimsRadius: 4
    ExtendIntoGroundDistance: 5
    MinimumWidth: 5
    MinimumArea: 100
    MaximumDepth: 0
    InvestigationTool: STICK
    ModificationTool: GOLDEN_SHOVEL
    Expiration:
      ChestClaimDays: 7
      UnusedClaimDays: 14
      AllClaims:
        DaysInactive: 60
        ExceptWhenOwnerHasTotalClaimBlocks: 10000
        ExceptWhenOwnerHasBonusClaimBlocks: 5000
      AutomaticNatureRestoration:
        SurvivalWorlds: false
    AllowTrappedInAdminClaims: false
    MaximumNumberOfClaimsPerPlayer: 0
    CreationRequiresWorldGuardBuildPermission: true
    VillagerTradingRequiresPermission: true
    CommandsRequiringAccessTrust: /sethome
    DeliverManuals: true
    ManualDeliveryDelaySeconds: 30
    RavagersBreakBlocks: true
    FireSpreadsInClaims: false
    FireDamagesInClaims: false
    LecternReadingRequiresAccessTrust: true
  Spam:
    Enabled: true
    LoginCooldownSeconds: 60
    LoginLogoutNotificationsPerMinute: 5
    ChatSlashCommands: /me;/global;/local
    WhisperSlashCommands: /tell;/pm;/r;/whisper;/msg
    WarningMessage: Please reduce your noise level.  Spammers will be banned. This
      is your warning!
    BanOffenders: true
    BanMessage: You have been banned for spam. Contact Jim to be unbanned
    AllowedIpAddresses: 1.2.3.4; 5.6.7.8
    DeathMessageCooldownSeconds: 120
    Logout Message Delay In Seconds: 0
  PvP:
    RulesEnabledInWorld:
      world: false
      world_nether: true
      world_the_end: true
    ProtectFreshSpawns: true
    PunishLogout: true
    CombatTimeoutSeconds: 15
    AllowCombatItemDrop: false
    BlockedSlashCommands: /home;/vanish;/spawn;/tpa
    ProtectPlayersInLandClaims:
      PlayerOwnedClaims: true
      AdministrativeClaims: true
      AdministrativeSubdivisions: true
    AllowLavaDumpingNearOtherPlayers:
      PvPWorlds: true
      NonPvPWorlds: false
    AllowFlintAndSteelNearOtherPlayers:
      PvPWorlds: true
      NonPvPWorlds: false
    ProtectPetsOutsideLandClaims: false
  Economy:
    ClaimBlocksPurchaseCost: 100.0
    ClaimBlocksSellValue: 50.0
  ProtectItemsDroppedOnDeath:
    PvPWorlds: false
    NonPvPWorlds: true
  BlockLandClaimExplosions: true
  BlockSurfaceCreeperExplosions: true
  BlockSurfaceOtherExplosions: true
  LimitSkyTrees: true
  LimitTreeGrowth: false
  CheckPistonMovement: true
  LimitPistonsToLandClaims: true
  FireSpreads: false
  FireDestroys: false
  AdminsGetWhispers: true
  AdminsGetSignNotifications: true
  SmartBan: true
  Mute New Players Using Banned Words: true
  MaxPlayersPerIpAddress: 3
  Siege:
    Worlds: []
    BreakableBlocks:
    - DIRT
    - GRASS_BLOCK
    - GRASS
    - FERN
    - DEAD_BUSH
    - COBBLESTONE
    - GRAVEL
    - SAND
    - GLASS
    - GLASS_PANE
    - OAK_PLANKS
    - SPRUCE_PLANKS
    - BIRCH_PLANKS
    - JUNGLE_PLANKS
    - ACACIA_PLANKS
    - DARK_OAK_PLANKS
    - WHITE_WOOL
    - ORANGE_WOOL
    - MAGENTA_WOOL
    - LIGHT_BLUE_WOOL
    - YELLOW_WOOL
    - LIME_WOOL
    - PINK_WOOL
    - GRAY_WOOL
    - LIGHT_GRAY_WOOL
    - CYAN_WOOL
    - PURPLE_WOOL
    - BLUE_WOOL
    - BROWN_WOOL
    - GREEN_WOOL
    - RED_WOOL
    - BLACK_WOOL
    - SNOW
    DoorsOpenDelayInSeconds: 300
    CooldownEndInMinutes: 60
  EndermenMoveBlocks: false
  SilverfishBreakBlocks: false
  CreaturesTrampleCrops: false
  RabbitsEatCrops: true
  HardModeZombiesBreakDoors: false
  Database:
    URL: ''
    UserName: ''
    Password: ''
  UseBanCommand: false
  BanCommandPattern: ban %name% %reason%
  Advanced:
    fixNegativeClaimblockAmounts: true
    ClaimExpirationCheckRate: 60
    OfflinePlayer_cache_days: 90
  Abridged Logs:
    Days To Keep: 7
    Included Entry Types:
      Social Activity: true
      Suspicious Activity: true
      Administrative Activity: false
      Debug: false
      Muted Chat Messages: false

```

### Plugin list (if applicable):

1. LuckPerms, 
2. CustomDeathMessages, 
3. AdvancedBan, 
4. WitherAC, 
5. ProtocolLib, 
6. dynmap, 
7. Essentials, 
8. EssentialsChat, 
9. Admin-GUI, 
10. ServerlistMOTD, 
11. Vault
12. GriefPrevention, 
13. SuperVanish

### Screenshot

![](https://cdn.discordapp.com/attachments/562168753635196958/727059038725996594/unknown.png)

### Debugging steps attempted

- [x] I attempted running only GriefPrevention on the server
  - [x] Issue still occurs even if GriefPrevention is the only plugin running
Note: Since to buy claim blocks there needs to be an economy plugin, I left Essentials & Vault running
- [x] I attempted testing for the issue on a new server
  - [x] Issue still occurs even after testing on a new server

**Comment by RoboMWM** [comment]
So uh, did these players have "negative" claim blocks before?

**Comment by Compositr** [comment]
No. It only happens after they attempt to buy blocks.

**Comment by RoboMWM** [comment]
Strange... 

Guess will need to review that logic. Does /acb work as intended?

**Comment by Compositr** [comment]
/acb 

returns
```
[GriefPrevention] Adjusted Aquamarine_Pearl's bonus claim blocks by 500. New total bonus blocks: 11061.

```
but now it says a random positive integer (after I did /acb)

![image](https://user-images.githubusercontent.com/43405050/86331243-29be4480-bc8c-11ea-914b-86a2c53923f3.png)
It says they have 11061 blocks when I use /acb, but they apparently don't have that many...

**Comment by RoboMWM** [comment]
use /claimslist to see total claimblock counts.

what you see there is the available claim blocks (accrued + bonus - claim areas). The output is stating the `bonus` value.

**Comment by RoboMWM** [comment]
So that being known now: When buying claim blocks, are those "bought" blocks correctly added to the player's `bonus` blocks?

**Comment by Compositr** [comment]
Yes, when buying blocks, they get added to the `bonus` blocks section, but sometimes the player's golden shovel tells them they have negative blocks.

This is what I get when doing /claimslist
```
 [GriefPrevention] 5000 blocks from play + 11561 bonus = 16061 total.

 [GriefPrevention] Claims:

 [GriefPrevention] world: x829, z311 (-3650 blocks)

 [GriefPrevention] world: x-1026, z-57 (-285 blocks)

 [GriefPrevention] world: x-1111, z1128 (-3240 blocks)

 [GriefPrevention] §eworld: x36, z-438 (-3420 blocks)

 [GriefPrevention] = 5966 blocks left to spend
```
Somehow their claims are negative blocks, don't know why...

**Comment by RoboMWM** [comment, marked as answer]
no, the claims aren't negative blocks. Instead it's showing you the calculation - as players claim, they "consume" their available claimblocks. What you see at the end is the remaining amount after calculating all the blocks used for all of the claims.

If there is ever a time that players have a negative number of claim blocks remaining, then somehow they're able to claim more than they are allocated to do so - this can happen via addons (e.g., GPRealEstate did not transfer claimblocks when transferring claims, an issue I fixed in my fork of the plugin) or if you do things where it's possible to temporarily have more bonus blocks available in some situations, such as giving a permission group an extra amount of blocks in a specific world.

**Comment by Jikoo** [comment]
Going to close this - assuming resolved due to lack of response.
